src/Adapter/Product/Update/CombinationStockUpdater.php

```diff
@@ -0,0 +1,144 @@
+
+namespace PrestaShop\PrestaShop\Adapter\Product\Update;
+
+use Combination;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\CombinationRepository;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\StockAvailableRepository;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Exception\CannotUpdateCombinationException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Exception\CombinationConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\ValueObject\CombinationId;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\Exception\StockAvailableNotFoundException;
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use PrestaShop\PrestaShop\Core\Stock\StockManager;
+use PrestaShopException;
+
+class CombinationStockUpdater
+{
+    /**
+     * @var StockAvailableRepository
+     */
+    private $stockAvailableRepository;
+
+    /**
+     * @var CombinationRepository
+     */
+    private $combinationRepository;
+
+    /**
+     * @var StockManager
+     */
+    private $stockManager;
+
+    /**
+     * @param StockAvailableRepository $stockAvailableRepository
+     * @param CombinationRepository $combinationRepository
+     * @param StockManager $stockManager
+     */
+    public function __construct(
+        StockAvailableRepository $stockAvailableRepository,
+        CombinationRepository $combinationRepository,
+        StockManager $stockManager
+    ) {
+        $this->stockAvailableRepository = $stockAvailableRepository;
+        $this->combinationRepository = $combinationRepository;
+        $this->stockManager = $stockManager;
+    }
+
+    /**
+     * @param Combination $combination
+     * @param array $propertiesToUpdate
+     * @param bool $addMovement
+     *
+     * @throws CombinationConstraintException
+     * @throws CoreException
+     * @throws StockAvailableNotFoundException
+     */
+    public function update(Combination $combination, array $propertiesToUpdate, bool $addMovement): void
+    {
+        $this->combinationRepository->partialUpdate(
+            $combination,
+            $propertiesToUpdate,
+            CannotUpdateCombinationException::FAILED_UPDATE_STOCK
+        );
+
+        // updating stockAvailable after Combination is important, because quantity validation is done in Combination update
+        if (in_array('quantity', $propertiesToUpdate)) {
+            $this->updateStockAvailable($combination, $addMovement);
+        }
+    }
+
+    /**
+     * @param Combination $combination
+     * @param bool $addMovement
+     *
+     * @throws CombinationConstraintException
+     * @throws CoreException
+     * @throws StockAvailableNotFoundException
+     */
+    private function updateStockAvailable(Combination $combination, bool $addMovement): void
+    {
+        $newQuantity = (int) $combination->quantity;
+        $stockAvailable = $this->stockAvailableRepository->getForCombination(new CombinationId((int) $combination->id));
```

Comment:
Can this variable be `null` or `[]` ?

Reply:
no,  `getForCombination` returns StockAvailable or throws exception

Reply:
Like for product at the Combination creation n associated StockAvailable entity is created by the legacy `add()` method so it should always be present It its not something went wrong but then the repository will throw an exception